cli/coverage_cmd.go

```diff
@@ -0,0 +1,174 @@
+package cli
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"go/build"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/tools/go/loader"
+
+	"github.com/kr/fs"
+
+	"sourcegraph.com/sourcegraph/go-flags"
+)
+
+func init() {
+	cliInit = append(cliInit, func(cli *flags.Command) {
+		_, err := cli.AddCommand("coverage",
+			"use a simple heuristic to check that srclib is outputting expected graph data",
+			`The coverage command acts as a sanity check to ensure that the srclib-go toolchain succeeds when it should and fails when it should not.`,
+			&coverageCmd,
+		)
+		if err != nil {
+			log.Fatal(err)
+		}
+	})
+}
+
+var cacheDir = ".srclib-cache"
+var gitDir = ".git"
+var unitFile = "GoPackage.unit.json"
+var depFile = "GoPackage.depresolve.json"
+var graphFile = "GoPackage.graph.json"
+
+type CoverageCmd struct {
+	w io.Writer
+}
+
+type Coverage struct {
+	Repo     *Repo
+	Warnings []BuildWarning
+}
+
+type BuildWarning struct {
+	Directory string
+	Warning   WarningType
+}
+
+type WarningType string
+
+const (
+	BuildSucceededSrclibFailed = "Build succeeded but Srclib outputs failed"
+	BuildFailedSrclibSucceeded = "Build failed but Srclib succeeded"
+)
+
+var coverageCmd CoverageCmd
+
+// Execute performs a sanity check on srclib builds for go repositories.
+// We iterate every directory and do a coverage check on every package.
+// The coverage heuristic is very rough currently, but makes the following assumptions:
+// - only golang coverage is checked
+// - if a directory can be imported (see build standard library package) and built
+//   (see loader standard library package), then there should be three files present
+//   in the corresponding directory under .srclib-cache: a unit file, a depresolve file,
+//   and a graph file.
+func (c *CoverageCmd) Execute(args []string) error {
+	if c.w == nil {
+		c.w = os.Stdout
+	}
+
+	lRepo, lRepoErr := OpenLocalRepo()
+	if lRepoErr != nil {
+		return lRepoErr
+	}
+
+	goPath := os.Getenv("GOPATH")
+	if goPath == "" {
+		return errors.New("GOPATH not set")
+	}
+
+	var importPath string
+	for _, p := range strings.Split(goPath, ":") {
+		if strings.Contains(lRepo.RootDir, p) {
+			importPath = strings.TrimPrefix(lRepo.RootDir, filepath.Join(p, "src")+"/")
```

Comment:
@alexsaveliev thanks for taking a look. I don't have a windows machine, would you be able to confirm this?